File: android/app/src/main/java/app/pulpe/android/domain/model/Transaction.kt
package app.pulpe.android.domain.model

import kotlinx.serialization.Serializable
import java.math.BigDecimal

/**
 * Transaction representing an actual financial operation
 */
@Serializable
data class Transaction(
    val id: String,
    val budgetId: String,
    val budgetLineId: String? = null,
    val name: String,
    val amount: Double,
    val kind: TransactionKind,
    val transactionDate: String,
    val category: String? = null,
    val checkedAt: String? = null,
    val createdAt: String,
    val updatedAt: String
) {
    val isChecked: Boolean
        get() = checkedAt != null

    val isAllocated: Boolean
        get() = budgetLineId != null

    val isFree: Boolean
        get() = budgetLineId == null

    val amountDecimal: BigDecimal
        get() = amount.toBigDecimal()

    fun toggled(): Transaction = copy(
        checkedAt = if (isChecked) null else java.time.Instant.now().toString(),
        updatedAt = java.time.Instant.now().toString()
    )
}

/**
 * Transaction creation DTO
 */
@Serializable
data class TransactionCreate(
    val budgetId: String,
    val budgetLineId: String? = null,
    val name: String,
    val amount: Double,
    val kind: TransactionKind,
    val transactionDate: String? = null,
    val category: String? = null,
    val checkedAt: String? = null
)

/**
 * Transaction update DTO
 */
@Serializable
data class TransactionUpdate(
    val name: String? = null,
    val amount: Double? = null,
    val kind: TransactionKind? = null,
    val transactionDate: String? = null,
    val category: String? = null
)

File: backend-dotnet/src/Pulpe.Api/Domain/Transaction/Transaction.cs
using Pulpe.Api.Domain.Common;

namespace Pulpe.Api.Domain.Transaction;

public sealed class Transaction
{
    public Guid Id { get; init; }
    public Guid BudgetId { get; init; }
    public Guid? BudgetLineId { get; init; }
    public string Name { get; init; } = string.Empty;
    public string Amount { get; set; } = string.Empty; // Encrypted
    public DateTimeOffset TransactionDate { get; init; }
    public string? Category { get; init; }
    public TransactionKind Kind { get; init; }
    public DateTimeOffset? CheckedAt { get; init; }
    public DateTimeOffset CreatedAt { get; init; }
    public DateTimeOffset UpdatedAt { get; init; }
}
